### PR TITLE
Changing translation process of permalink and slugs in models SP-56

### PR DESCRIPTION
**Pull request by kpokorski**
Changing a way of saving translations in taxon.rb and product.rb models in order to enable the proper generation of permalinks and slugs. 
**This PR is dependable on [this](https://github.com/spree/spree_backend/pull/262) spree_backend PR**

### CODE REVIEW

File: core/app/models/spree/product.rb [hunk 1]
```diff
@@ -42,11 +42,24 @@ class Product < Spree::Base
     translates(*TRANSLATABLE_FIELDS)
 
     self::Translation.class_eval do
+      before_update :set_slug
       acts_as_paranoid
       # deleted translation values also need to be accessible for index views listing deleted resources
       default_scope { unscope(where: :deleted_at) }
+      def set_slug
+        if self.name.blank? & self.slug.blank?
```
Comment (kshalot) [inline core/app/models/spree/product.rb:50]: You should not use bit operators to do boolean logic like this. Stick to `&&`.

File: core/app/models/spree/taxon.rb [hunk 2]
```diff
@@ -62,13 +62,15 @@ class Taxon < Spree::Base
...
       def set_permalink
         parent = translated_model.parent
         name_with_fallback = name || translated_model.name
 
+        if self.name.blank? & self.permalink.blank?
+          self.permalink = translated_model.name.to_url
+        end
         if parent.present?
           self.permalink = [parent.permalink, (permalink.blank? ? name_with_fallback.to_url : permalink.split('/').last)].join('/')
         else
```
Comment (kshalot) [inline core/app/models/spree/taxon.rb:76]: You first set permalink to:
```rb
translated_model.name.to_url
```
then to
```rb
[parent.permalink, (permalink.blank? ? name_with_fallback.to_url : permalink.split('/').last)].join('/')
```

The second will overwrite the other. Is this intentional? Is this missing and `elsif`

File: core/app/models/spree/product.rb [hunk 3]
```diff
@@ -42,11 +42,24 @@ class Product < Spree::Base
...
+            self.slug = translated_model.translations.find_by(locale: translated_model.stores.first.default_locale).name.to_url
+        else
+          if self.slug.blank?
+            self.slug = self.name.to_url
+          else
+            self.slug = self.slug.to_url
```
Comment (kshalot) [inline core/app/models/spree/product.rb:56]: What does this do? Sets the slug to slug - what's the exact use case?

**Review by kshalot — CHANGES_REQUESTED** [review]

File: core/app/models/spree/product.rb (hunk 1 above)
Comment (kpokorski) [reply core/app/models/spree/product.rb]: ups, typo

File: core/app/models/spree/product.rb (hunk 3 above)
Comment (kpokorski) [reply core/app/models/spree/product.rb]: actually, the only use case here is making sure, the slug is in slug format.

File: core/spec/models/spree/taxon_spec.rb [hunk 4]
```diff
@@ -189,7 +189,10 @@
     let(:taxon) { create(:taxon, taxonomy: taxonomy, permalink: 'test_slug_en') }
     let!(:taxon_translation_fr) { taxon.translations.create(slug: 'test_slug_fr', locale: 'fr') }
 
-    before { Spree::Locales::SetFallbackLocaleForStore.new.call(store: store) }
+    before do
+      @root_taxon = taxonomy.taxons.find_by(parent_id: nil)
```
Comment (rafalcymerys) [inline core/spec/models/spree/taxon_spec.rb:193]: Why don't we put that into a `let` instead?

**Review by rafalcymerys — APPROVED** [review]

**Pull request merged**